internal/remotes/tmc.go

```diff
@@ -0,0 +1,278 @@
+package remotes
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/web-of-things-open-source/tm-catalog-cli/internal/app/http/server"
+	"github.com/web-of-things-open-source/tm-catalog-cli/internal/model"
+	"github.com/web-of-things-open-source/tm-catalog-cli/internal/utils"
+)
+
+const (
+	headerContentType = "Content-Type"
+	mimeJSON          = "application/json"
+)
+
+// TmcRemote implements a Remote TM repository backed by an instance of TM catalog REST API server
+type TmcRemote struct {
+	baseHttpRemote
+}
+
+func NewTmcRemote(config map[string]any, spec RepoSpec) (*TmcRemote, error) {
+	base, err := newBaseHttpRemote(config, spec)
+	if err != nil {
+		return nil, err
+	}
+	r := &TmcRemote{baseHttpRemote: base}
+	return r, nil
+}
+
+func (t TmcRemote) Push(id model.TMID, raw []byte) error {
+	reqUrl := t.parsedRoot.JoinPath("thing-models")
+	req, err := http.NewRequest(http.MethodPost, reqUrl.String(), bytes.NewBuffer(raw))
+	if err != nil {
+		return err
+	}
+	req.Header.Add(headerContentType, mimeJSON)
+	resp, err := doHttp(req, t.auth)
+	if err != nil {
+		return err
+	}
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	switch resp.StatusCode {
+	case http.StatusCreated:
+		return nil
+	case http.StatusConflict, http.StatusInternalServerError, http.StatusBadRequest:
+		var e server.ErrorResponse
+		err = json.Unmarshal(b, &e)
+		if err != nil {
+			return err
+		}
+		detail := e.Title
+		if e.Detail != nil {
+			detail = *e.Detail
+		}
+		switch resp.StatusCode {
+		case http.StatusConflict:
+			err := &ErrTMExists{}
+			err.FromString(detail)
+			return err
+		case http.StatusInternalServerError, http.StatusBadRequest:
+			return errors.New(detail)
+		default:
+			return errors.New("unexpected status not handled correctly")
+		}
+	default:
+		return errors.New(fmt.Sprintf("received unexpected HTTP response from remote TM catalog: %s", resp.Status))
+	}
+}
+
+func (t TmcRemote) Spec() RepoSpec {
+	return t.spec
+}
+func (t TmcRemote) Fetch(id string) (string, []byte, error) {
+	reqUrl := t.parsedRoot.JoinPath("thing-models", id)
+	return fetchTM(reqUrl.String(), t.auth)
+}
+
+func (t TmcRemote) UpdateToc(...string) error {
+	return nil // ignore request to update toc as toc updates are presumed to be handled by the underlying remote
+}
+
+func (t TmcRemote) List(search *model.SearchParams) (model.SearchResult, error) {
```

Comment:
Maybe its worth to discuss, as we do the filter stuff on server side:

Lets assume we have a hosted TMC remote, with tmc-catalog-cli  version 1.0.0 
Then we use a tmc-catalog-cli version 2.0.0, that has more filter that can be set.  
This would lead to a wrong result, as the new filter would be ignored on the hosted TMC remote.  
Of course, if there would be Rest API breaking changes, we would have even more problems.